Change Handling of Some Mistypings and Oversized Values

rmarrowstone opened the pull request:
This commit has a few changes to edges in handling of mistyped or
oversized values. Each change either brings this impl in line with
Ion Hive or makes it stricter.

* IonInts are coerced to Decimals (as Ion Hive does)
* Timestamps are truncated when overprecise or being coerced to Dates
  (as Ion Hive does, see note below)
* Decimals with more whole digits than fit is now an error
  (oddly this is not an error in Ion Hive)
* TrinoExceptions are thrown for a few cases

Fairly exhaustive test cases were added for each of the changes.

A note about rounding vs truncating Timestamps: the SQL standard
says that it's implementation defined what to do here. For me,
the fact that a timestamp of 2023-12-31 23:59:59.999999Z would
be rounded into 2024 is just too surprising to consider correct.
So, given that and that the legacy behavior is to truncate, that
seems correct to me.

rmarrowstone committed:
Coerce Ints to Decimals, Handle Oversized Numbers

This commit has a few changes to edges in handling of mistyped or
oversized values. Each change either brings this impl in line with
Ion Hive or makes it stricter.

* IonInts are coerced to Decimals (as Ion Hive does)
* Decimals with more whole digits than fit is now an error
  (oddly this is not an error in Ion Hive)
* TrinoExceptions are thrown for a few cases

Fairly exhaustive test cases were added for each of the changes.

rmarrowstone commented:
I'm going to revisit the Timestamp rounding after getting input from the Athena folks.

Apparently the answer is to use `Timestamps.roundDiv` which will round in the Trino OSS code, but truncate for Athena.

rmarrowstone commented:
> I'm going to revisit the Timestamp rounding after getting input from the Athena folks.
> 
> Apparently the answer is to use `Timestamps.roundDiv` which will round in the Trino OSS code, but truncate for Athena.

Updated to do so. I pulled the plug past nanos, that just seemed sufficient to me.

linlin-s commented on lib/trino-hive-formats/src/main/java/io/trino/hive/formats/ion/IonDecoderFactory.java:
```diff
@@ -225,68 +231,83 @@ public void decode(IonReader ionReader, BlockBuilder blockBuilder)
-            BigintType.BIGINT.writeLong(blockBuilder, ionReader.longValue());
+            BigintType.BIGINT.writeLong(blockBuilder, readLong(ionReader));
+
+    private static long readLong(IonReader ionReader)
+    {
+        try {
+            return ionReader.longValue();
+        }
+        catch (IonException e) {
+            throw new TrinoException(StandardErrorCode.GENERIC_USER_ERROR, e.getMessage());
+        }
+    }
```
For consistency, should we consider implementing similar error handling for other Ion data type reading methods? For example, `ionReader.doubleValue()` or` ionReader.booleanValue()`?

rmarrowstone replied:
This is different than those. With those, if the IonType is as expected you are not going to get an Exception from calling them. In this case, if the IonInt is larger than a `long` then you get an IonException here.

linlin-s commented on lib/trino-hive-formats/src/main/java/io/trino/hive/formats/ion/IonDecoderFactory.java:
```diff
@@ -225,68 +231,83 @@ public void decode(IonReader ionReader, BlockBuilder blockBuilder)
     private static final BlockDecoder realDecoder = (ionReader, blockBuilder) -> {
         double readValue = ionReader.doubleValue();
         if (readValue == (float) readValue) {
             RealType.REAL.writeFloat(blockBuilder, (float) ionReader.doubleValue());
```
Nit Pick: Consider creating variable for `(float) ionReader.doubleValue()` and reuse it.

rmarrowstone replied:
yea i could reuse the already defined `readValue` in the block call.

rmarrowstone replied:
Addressed this.

linlin-s commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:
```diff
@@ -8754,6 +8767,14 @@ private void testTimestampPrecisionWrites(Session session, String tableName, BiC
+        assertQuery(withTimestampPrecision(session, HiveTimestampPrecision.MILLISECONDS), sql, "VALUES ('2019-02-03 18:30:00.556')");
+
+        populateData.accept("2019-02-03 18:30:00.444500", HiveTimestampPrecision.MICROSECONDS);
+        assertQuery(withTimestampPrecision(session, HiveTimestampPrecision.MILLISECONDS), sql, "VALUES ('2019-02-03 18:30:00.445')");
```
For the rounding mode `HALF_UP`:

> Round towards the "nearest neighbor". If both neighbors are equidistant, round up. Behaves as `RoundingMode.UP` if the discarded fraction is ≥ 0.5. Otherwise, behaves as `RoundingMode.DOWN`

Should we consider replacing the second case with a `RoundingMode.DOWN` case? For example:
```
populateData.accept("2019-02-03 18:30:00.444300", HiveTimestampPrecision.MICROSECONDS);
assertQuery(withTimestampPrecision(session, HiveTimestampPrecision.MILLISECONDS), sql, "VALUES ('2019-02-03 18:30:00.444')");
Comment

```

rmarrowstone replied:
Sorry I meant to remove this from the diff. Now that I know the behavior for timestamp rounding is different in Athena I don't want to pile on to what I'm sure is a source of merge conflicts. So I will remove this from the diff and stay in my lane. :)

linlin-s replied:
No worries, I left this comment before the new change of the Timestamp. Removing this from the diff sounds good to me.

rmarrowstone replied:
Removed.

rmarrowstone commented:
> > I'm going to revisit the Timestamp rounding after getting input from the Athena folks.
> > Apparently the answer is to use `Timestamps.roundDiv` which will round in the Trino OSS code, but truncate for Athena.
> 
> Updated to do so. I pulled the plug past nanos, that just seemed sufficient to me.

I couldn't let the fractional picos part go. Updated again to cover picos. Done now.

rmarrowstone committed:
Ensure Timestamps are Rounded Consistently

This change makes it so that Ion Timestamps are rounded using
Timestamps.round(). This makes it so that Timestamp down casting works
consistently with the other formats.

I considered using DecodedTimestamp and the TrinoTimestampEncoders
but those didn't cover picos. So this code uses the same pattern they
do and the Timestamps.round(). This code simply ignores anything after
picos.
